### ISSUE TITLE
`cargo doc --open` should start a web server to enable keyboard accessibility

### ISSUE THREAD

**Issue by kbdluv** [rust-lang/cargo#4966]
Running `cargo doc --open` currently opens the file in the browser with the `file://` protocol. 

In this protocol, web extensions like **Vimium** don't have access to the page and can't provide keyboard shortcuts. 

It would be nice if running `cargo doc --open` would start a web server like [http](https://crates.io/crates/https) then navigate to the generated index file. I've tested that and Vimium works normally. Thank you!

**Comment by steveklabnik** [comment]
This has historically been controversial. I'm pro, but there are a lot of people against.

**Comment by kbdluv** [comment]
What are the con arguments?

**Comment by est31** [comment]
I've just installed [Vimium](https://addons.mozilla.org/firefox/addon/vimium-ff/) on Firefox, it works great. So maybe install Firefox :).

**Comment by chabapok** [comment]
As far as I understand, it's browser or vimium problem.

But if we do this, it makes sense to use separate key, for example: `cargo doc --open_with_embedded_webserver`

**Comment by est31** [comment]
Yeah it is a Chrome bug. In general, Chrome cares only very little about the `file://` protocol: https://bugs.chromium.org/p/chromium/issues/detail?id=47416

But why should firefox users being made suffer? Firefox works great, but http://localhost is far more inconvenient than file protocol based services.

**Comment by qmx** [comment]
having a built-in server would be really useful for those that usually work inside a ssh session/vm and use a browser outside the shell session.

**Comment by pravic** [comment]
> for those that usually work inside a ssh session/vm

Why trying to squash the universe inside one tool? Let cargo doc *generate* documentation, let another tool to host it.

**Comment by qmx** [comment]
I'm hacking together a `cargo docserver` for this, hopefully it gets somewhere 🙏

**Comment by qmx** [comment]
just pushed https://crates.io/crates/cargo-docserver, hopefully that's useful for someone besides me

**Comment by wez** [comment]
@qmx it works, thanks!

**Comment by xypnox** [comment]
Probably if it is an edge case scenario, you can use https://www.npmjs.com/package/serve to serve your docs to a localhost and then open that in chrome(ium).

This is much simpler solution than building a server inside cargo itself.

**Comment by qlcom** [comment]
> This has historically been controversial. I'm pro, but there are a lot of people against.

Then why you use a web browser.

**Comment by qlcom** [comment]
> Probably if it is an edge case scenario, you can use https://www.npmjs.com/package/serve to serve your docs to a localhost and then open that in chrome(ium).
> 
> This is much simpler solution than building a server inside cargo itself.

How can this be simple? You need nodejs to run it. Have you lost your mind?

**Comment by xypnox** [comment]
> How can this be simple? You need nodejs to run it. Have you lost your mind?

No I haven't. What I meant with the suggestion was that cargo is not meant to be a server. Including an http server just to view docs on localhost doesn't make much sense when there are libraries such as [cargo-docserver](https://crates.io/crates/cargo-docserver).

I was just suggesting a simple solution in terms of the time it would take for someone to use it. Granted it is simple only for those who already have node installed, but it is a simple solution nonetheless.

**Comment by frehberg** [comment]
Web-browser are gettting more strict regarding file access. As the rustdoc is organized in folders hierarchically, within each document the relative location of the static.files/ is derived reading the following attribute 
```
var rootPath = document
         .getElementById('rustdoc-vars')
         .attributes['data-root-path']
         .value;
```
Due to this fact the import paths for further scripts/files are constructed dynamically using rootPath as prefix. When using the file-protocol to read the rustdoc, these dynamic imports will be blocked by the browser, see CORS error message of Chrome below.
```
   await import(rootPath + "static-files/....js");
```
![image](https://user-images.githubusercontent.com/4435830/224663673-21e39cd5-6489-4026-b4f6-5fed5b234c19.png)

This means, that certain styles and features can not be loaded form folder static.files and the rustdocs will appear different compared to using the http-protocol.

The solution might be to turn to static import paths. This could be achieved by flattening the folder structure of the rustdocs, using a file set for each crate of folder depth=1 only, so that a static string could be use for the import paths always.
```
import x from "../static.files/...js";
```
 
Using static import paths, the CORS issue would no longer exist and the file-protocol could be used the same way as http-protocol. ( AFAICS)

PS: other rustdocs add-ons such as aquamarine must evaluate rootPath value as well, right now. Using a flattened folder structure (depth=1), the integration of these tools would be simplified.

**Comment by LechintanTudor** [comment]
This change would be very useful to me as I run Firefox in a Flatpak container which only has access to `~/Downloads`.

**Comment by peterramaldes** [comment]
Yea, I tried use here too with `python3 -m http.server` and no success :(

![image](https://github.com/rust-lang/cargo/assets/3129159/489e79ff-fdd4-4c34-a6ca-078091bc3cf7)

**Comment by SteveLauC** [comment]
> just pushed https://crates.io/crates/cargo-docserver, hopefully that's useful for someone besides me

Hi @qmx, may I ask why this [repo is archived](https://github.com/qmx/cargo-docserver)?

**Comment by SamB** [comment]
`cargo doc --open` also doesn't work from Termux on my phone. Termux tries to open some `content://` URL in a browser but it doesn't work, and when I attempt to manually open the docs in a browser, Chrome doesn't even seem to recognize `file://` URLs, and Firefox appears to treat them as aliases for `about:blank`. (I was actually _expecting_ to just get a read error because `/data/data/com.termux/files/home` can't be read by other apps.)

**Comment by jdavisp3** [comment]
My solution is using `--target-dir` to put the docs in a location that Chrome is happy with (~/Documents).

**Comment by qmx** [comment]
> > just pushed https://crates.io/crates/cargo-docserver, hopefully that's useful for someone besides me
> 
> 
> 
> Hi @qmx, may I ask why this [repo is archived](https://github.com/qmx/cargo-docserver)?

Reason is simple: I stopped maintaining it!

**Comment by ehuss** [comment]
The cargo team briefly discussed this (and I brought it up with the rustdoc team at https://rust-lang.zulipchat.com/#narrow/channel/266220-t-rustdoc/topic/local.20server/with/522152745 as well). We didn't have any specific blocking concerns other than some minor concerns. Some considerations might be:
- The weight of the dependencies necessary to support this.
- The maintenance burden on the cargo team.
    - See https://github.com/rust-lang/mdBook/issues/?q=is%3Aissue%20state%3Aopen%20label%3ACommand-serve for similar burden that happens with mdbook.
- Avoid scope creep.
- Making it clear this should not be used in production.
- Possible coordination with the rustup team for supporting a similar feature for `rustup doc`.

**Comment by epage** [comment]
Design wise, I've been considering whether we should have interactive modes for our commands which would include include watch support (rebuilding on changes).  I wonder if we should tie running a server into that.

**Comment by rakshith-ravi** [comment]
Doesn't seem to be in the previous conversation (apologies if I missed it) - when working with a DevContainer or a remote machine, it's not always easy to get access to the files on the host machine after cargo docs have been generated. LOT of files to be downloaded.

`cargo doc --open` would require just a simple port-forward (available in most remove dev environments) that can allow us to view the docs. Especially useful in macro-filled projects to see if the types are getting generated correctly.

**Comment by rakshith-ravi** [comment]
@qmx would you be open to transferring ownership of the crate? Happy to maintain it going forward

**Comment by epage** [comment]
This was marked needs-mentor but in the meeting notes and the summary here, this sounds like there are still a lot of open questions around this so I switched to needs-design.

Implementation wise, we have
- https://crates.io/crates/cargo-docserver
- pulling out mdbook's serve logic
- [file-serve](https://docs.rs/file-serve/latest/file_serve/) which I extracted from a SSG I maintain and attempts to stay light weight to build and has "find a free port" mode

It might also be good to survey other tools with a serve functionality, e.g. mdbook and my SSG allow overriding the hostname and port.  Is that required for an MVP?  Does it need to be on the command-line or could it be in config (which can also be set via env / command-line).
